Fix random failure on system test with ajax

mtsmfm opened the pull request:
### Summary

If application has ajax, browser may begin request after rollback.
It breaks tests randomly.

https://github.com/mtsmfm/rails-system-test-example is a demo application.

#### ajax request

https://github.com/mtsmfm/rails-system-test-example/blob/fail/app/assets/javascripts/application.js#L15

#### system test (scaffold)

https://github.com/mtsmfm/rails-system-test-example/blob/fail/test/system/posts_test.rb#L11-L20

#### build results

- https://circleci.com/gh/mtsmfm/rails-system-test-example/11 (failed)
- https://circleci.com/gh/mtsmfm/rails-system-test-example/12 (fixed by using this branch)

### Other Information

@iangreenleaf wrote great article about this problem :star2:

http://technotes.iangreenleaf.com/posts/the-one-true-guide-to-database-transactions-with-capybara.html

It helped me to notice what happened in system test, thanks!

In the article, suggesting solution is wait until ajax finished by `evaluate_script` but I think we can guard from ajax request after teardown by ~~inserting rack middleware~~ call `Capybara.reset_session!` before rollback.

iangreenleaf commented:
This is a cool solution! It doesn't cover 100% of the causes of ajax DB contention, but requests after teardown are definitely the most common cause. And it's a simple solution that doesn't overstep Rails' responsibilities and doesn't have to try to monkey-patch various JS frameworks.

iangreenleaf commented on actionpack/lib/action_dispatch/system_test_case.rb:
```diff
@@ -84,9 +84,29 @@ class SystemTestCase < IntegrationTest
+          if ActionDispatch::SystemTestCase.running?
+            @app.call(env)
+          else
+            [200, { "Content-Type" => "text/plain" }, ["OK"]]
```
I wonder if Rack has a way to just drop the connection? Or failing that, maybe a 500 status would be better. It feels a little weird to return a fake success status.

mtsmfm committed:
Fix random failure on system test with ajax

If application has ajax, browser may begin request after rollback.
`teardown_fixtures` will be called after `super` on `after_teardown`
so we must call `Capybara.reset_sessions!` before `super`

https://github.com/rails/rails/blob/b61a56541aecd7ac685d4f19d943177a3f1b465a/activerecord/lib/active_record/fixtures.rb#L857

mtsmfm commented:
I found very simple solution.
Call `Capybara.reset_session!` before teardown.

https://circleci.com/gh/mtsmfm/rails-system-test-example/13

We don't have to insert middleware :innocent:

mtsmfm commented:
> It doesn't cover 100% of the causes of ajax DB contention

@iangreenleaf

Could you tell me more detail about another failing case :question: 
I'm very happy if we can fix that case but I don't know about another case :dizzy_face:

eileencodes commented:
I definitely prefer a simpler solution, but I'm unsure if putting this in `before_teardown` will muck with any settings someone puts in a `teardown` block. For example, if for some reason the user also wanted to take a screenshot in the `teardown` block it will be a screenshot of a white screen because you cannot clear sessions before taking a screenshot.

mtsmfm commented:
> For example, if for some reason the user also wanted to take a screenshot in the teardown block it will be a screenshot of a white screen because you cannot clear sessions before taking a screenshot.

:ok: 

Then, let's call `Capybara.reset_session!` before `super` because [`teardown_fixtures` called after `super` on `after_teardown`]( https://github.com/rails/rails/blob/b61a56541aecd7ac685d4f19d943177a3f1b465a/activerecord/lib/active_record/fixtures.rb#L857).

eileencodes commented:
Ok can you add a test that verified order so if someone moves Capbyara.reset_sessions in a future PR it fails? Really appreciate you working on this!

iangreenleaf commented:
@mtsmfm for example, the second most common cause I see is tests mixing browser interactions with direct DB access, like this:

```ruby
def test_my_page
  page.click 'Ajax Button'
  MyModel.create! name: "new model"
  # ...
end
```

However, I don't think Rails alone can do anything to protect from that. My gem offers a `wait_for_ajax` helper but it must plug into the JS framework to know about pending requests.

mtsmfm commented:
@iangreenleaf Thank you for your explanation!
Probably https://github.com/rails/rails/pull/28083 solved mixing problem because we have `lock` for shared connection now.

I experimented and couldn't got any failures:

- commit: https://github.com/mtsmfm/rails-system-test-example/commit/23e6ed15291accaf257ff22e1f6236c0cc192fbb
- test result: https://circleci.com/gh/mtsmfm/rails-system-test-example/18

@eileencodes Added a test and force pushed!

eileencodes commented on railties/test/application/test_runner_test.rb:
```diff
@@ -576,6 +576,34 @@ def test_warnings
         capture(:stderr) { run_test_command("test/models/warnings_test.rb -w") })
     end
 
+    def test_reset_sessions_before_rollback
```
Can you change this to be `test_reset_sessions_before_rollback_on_system_tests`? I think that's a better example of what this test is testing since it only applies to system tests.

mtsmfm commented:
@eileencodes Renamed and force pushed!

eileencodes commented:
Thank you!
